src/crypto/hash.cpp

```diff
@@ -12,45 +12,108 @@ extern "C"
 
 using namespace std;
 
-void crypto::Sha256Hash::mbedtls_sha256(const CBuffer& data, uint8_t* h)
+namespace crypto
 {
-  mbedtls_sha256_context ctx;
-  mbedtls_sha256_starts_ret(&ctx, 0);
+  constexpr auto block_size = 64u; // No way to ask evercrypt for this
 
-  mbedtls_sha256_update_ret(&ctx, data.p, data.rawSize());
+  void Sha256Hash::mbedtls_sha256(const CBuffer& data, uint8_t* h)
+  {
+    mbedtls_sha256_context ctx;
+    mbedtls_sha256_starts_ret(&ctx, 0);
 
-  mbedtls_sha256_finish_ret(&ctx, h);
-  mbedtls_sha256_free(&ctx);
-}
+    mbedtls_sha256_update_ret(&ctx, data.p, data.rawSize());
 
-void crypto::Sha256Hash::evercrypt_sha256(const CBuffer& data, uint8_t* h)
-{
-  EverCrypt_Hash_state_s* state =
-    EverCrypt_Hash_create(Spec_Hash_Definitions_SHA2_256);
-  EverCrypt_Hash_init(state);
+    mbedtls_sha256_finish_ret(&ctx, h);
+    mbedtls_sha256_free(&ctx);
+  }
 
-  constexpr auto block_size = 64u; // No way to ask evercrypt for this
+  void update_evercrypt_sha256(
+    const CBuffer& data, EverCrypt_Hash_state_s* state)
+  {
+    const auto data_begin = const_cast<uint8_t*>(data.p);
+    const auto size = data.rawSize();
+    const auto full_blocks = size / block_size;
 
-  const auto data_begin = const_cast<uint8_t*>(data.p);
-  const auto size = data.rawSize();
-  const auto full_blocks = size / block_size;
+    const auto full_blocks_size = full_blocks * block_size;
+    const auto full_blocks_end = data_begin + full_blocks_size;
 
-  const auto full_blocks_size = full_blocks * block_size;
-  const auto full_blocks_end = data_begin + full_blocks_size;
+    // update_multi takes complete blocks
+    EverCrypt_Hash_update_multi(state, data_begin, full_blocks_size);
 
-  // update_multi takes complete blocks
-  EverCrypt_Hash_update_multi(state, data_begin, full_blocks_size);
+    // update_last takes start of last chunk (NOT a full block!), and _total
+    // size_
+    EverCrypt_Hash_update_last(state, full_blocks_end, size);
+  }
 
-  // update_last takes start of last chunk (NOT a full block!), and _total size_
-  EverCrypt_Hash_update_last(state, full_blocks_end, size);
+  void Sha256Hash::evercrypt_sha256(const CBuffer& data, uint8_t* h)
```

Comment:
Could this function use the new `CSha256Impl` class to avoid calling the Evercrypt code from two places?